### ISSUE TITLE
Should editor consider code actions that are not listed in `codeActionKinds`?

### ISSUE THREAD

**Issue by rchl** [microsoft/language-server-protocol#1629]
Server can specify code actions that it supports through the `codeActionProvider.codeActionKinds` server capability. The spec says this about it:

```
	/**
	 * CodeActionKinds that this server may return.
	 *
	 * The list of kinds may be generic, such as `CodeActionKind.Refactor`,
	 * or the server may list out every specific kind they provide.
	 */
	codeActionKinds?: CodeActionKind[];
```

My questions are:
 - Should editor only allow code actions specified in `codeActionProvider.codeActionKinds` to be run as part of the "code actions on save" functionality? I know that "Code actions on save" are not really part of the spec but maybe you have opinion as someone who works with the "reference implementation" (VSCode).
 - Besides "code actions on save", what was the thinking behind introducing this capability? What practical effect should it have on the client from the LSP perspective?

**Comment by dbaeumer** [comment]
Regarding

> Besides "code actions on save", what was the thinking behind introducing this capability? What practical effect should it have on the client from the LSP perspective?

To avoid asking the server for code actions if they are not providing code actions for a specific kind.

For code actions on save: the client will ask with a specific kind for code actions on save. The kind is `CodeActionKind.SourceFixAll`

Hope that answers your questions.

**Comment by rchl** [comment]
I believe the first part of the question wasn't answered:

> Should editor only allow code actions specified in `codeActionProvider.codeActionKinds` to be run as part of the "code actions on save" functionality?

**Comment by dbaeumer** [comment]
When  the editor is asking for code actions on save the code action kind should be `CodeActionKind.Source.FixAll` or a subkind of it. If the server has not registered for that kind the editor shouldn't ask the server for code actions on save (see https://github.com/microsoft/vscode-eslint/blob/main/server/src/eslintServer.ts#L224 as an example)

**Comment by rchl** [comment]
Would you add something to that effect to the spec?

Note that VSCode doesn't follow that currently and will ask for code actions on save even if those are not registered through `codeActionProvider.codeActionKinds` server capability.

**Comment by dbaeumer** [comment]
> Note that VSCode doesn't follow that currently and will ask for code actions on save even if those are not registered through codeActionProvider.codeActionKinds server capability.

Good point and this is not only VS Code. We need to keep this for backwards compatibility since code actions on save where there before we add the kind registration.
